websocket_server/Services/ClusterService.js

```diff
@@ -0,0 +1,353 @@
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+/* eslint-disable no-console */
+
+export default class ClusterService {
+
+	constructor({
+		redisClient,
+		roomStateStore,
+		nodeId,
+		sessionTtl,
+		heartbeatTtlMs = 15000,
+		heartbeatKeyPrefix = 'node:',
+	} = {}) {
+		this.redisClient = redisClient
+		this.roomStateStore = roomStateStore
+		this.nodeId = nodeId
+		this.sessionTtl = sessionTtl
+		this.heartbeatTtlMs = heartbeatTtlMs
+		this.heartbeatKeyPrefix = heartbeatKeyPrefix
+		this.onSweep = null
+
+		this.heartbeatKey = this.buildHeartbeatKey(this.nodeId)
+		this.heartbeatInterval = null
+		this.stateSweepInterval = null
+	}
+
+	shouldUseRedis() {
+		return !!this.redisClient
+	}
+
+	getPresentationKey(roomId) {
+		return `room:${roomId}:presentation`
+	}
+
+	getRecordingKey(roomId) {
+		return `room:${roomId}:recording`
+	}
+
+	getRoomSyncerKey(roomId) {
+		return `room:${roomId}:syncer`
+	}
+
+	extractRoomId(key, suffix) {
+		const parts = key.split(':')
+		if (parts.length < 3) return null
+		if (parts[0] !== 'room') return null
+		if (parts[parts.length - 1] !== suffix) return null
+		return parts.slice(1, -1).join(':')
+	}
+
+	buildHeartbeatKey(nodeId) {
+		return `${this.heartbeatKeyPrefix}${nodeId}:heartbeat`
+	}
+
+	isClientClosedError(error) {
+		return error?.name === 'ClientClosedError' || error?.message?.includes('The client is closed')
+	}
+
+	setSweepHandler(handler) {
+		this.onSweep = handler
+	}
+
+	async start() {
+		await this.startHeartbeat()
+		this.startStateSweeper()
+	}
+
+	async stop() {
+		await this.stopStateSweeper()
+		await this.stopHeartbeat()
+	}
+
+	async startHeartbeat() {
+		if (!this.shouldUseRedis() || this.heartbeatInterval) {
+			return
+		}
+
+		const ttlSeconds = Math.max(1, Math.ceil(this.heartbeatTtlMs / 1000))
+		const writeHeartbeat = async () => {
+			try {
+				await this.redisClient.set(this.heartbeatKey, 'alive', { EX: ttlSeconds })
+			} catch (error) {
+				if (this.isClientClosedError(error)) {
+					return
+				}
+				console.error('Failed to write heartbeat:', error)
+			}
+		}
+
+		await writeHeartbeat()
+		this.heartbeatInterval = setInterval(writeHeartbeat, Math.max(1000, this.heartbeatTtlMs / 3))
+	}
+
+	async stopHeartbeat() {
+		if (this.heartbeatInterval) {
+			clearInterval(this.heartbeatInterval)
+			this.heartbeatInterval = null
+		}
+		if (this.shouldUseRedis()) {
+			try {
+				await this.redisClient.del(this.heartbeatKey)
+			} catch (error) {
+				if (this.isClientClosedError(error)) {
+					return
+				}
+				console.error('Failed to remove heartbeat key:', error)
+			}
+		}
+	}
+
+	startStateSweeper() {
+		if (!this.shouldUseRedis() || this.stateSweepInterval) {
+			return
+		}
+		const intervalMs = Math.max(2000, Math.floor(this.heartbeatTtlMs / 2))
+		this.stateSweepInterval = setInterval(() => {
+			this.runSweep().catch((error) => {
+				console.error('Failed to sweep cluster state:', error)
+			})
+		}, intervalMs)
```

Comment:
This should only happen on one node, otherwise this may lead to quite some performance impact on the redis cluster with more nodes. We could have some locking so only one node runs it.